src/cmd/linuxkit/pkglib/build.go

```diff
@@ -231,15 +231,18 @@ func (p Pkg) Build(bos ...BuildOpt) error {
 
 	skipBuild := bo.skipBuild
 	if !bo.force {
+		notFound := false
 		fmt.Fprintf(writer, "checking for %s in local cache, fallback to remote registry...\n", ref)
-		if _, err := c.ImagePull(&ref, "", arch, false); err == nil {
-			fmt.Fprintf(writer, "%s found or pulled\n", ref)
-			skipBuild = true
-		} else {
-			fmt.Fprintf(writer, "%s not found\n", ref)
+		for _, platform := range bo.platforms {
+			if _, err := c.ImagePull(&ref, "", platform.Architecture, false); err == nil {
```

Comment:
Is this going to work? We repeat `ImagePull()` each time with the same `ref` but a different architecture. Won't that lead to the `ref` pointing only to the last one in the list?

This might be easier if you just changed `ImagePull()` to support a list of architectures, so that you can keep this section unchanged, but have:

```go
c.ImagePull(&ref, "", bo.platforms, false)
```

Reply:
I use `p.IndexWrite(ref, im.Manifests...)` inside `ImagePull`, so it will store all manifests, cannot see any problems with updating to the last one.
Changing of ImagePull signature will affect another functions and seems we still need to know do we have all arches pulled or only partial of them, so return object of the function will also changed. I cannot see any great benefit of the changing.

Reply:
> I use p.IndexWrite(ref, im.Manifests...) inside ImagePull, so it will store all manifests

Correct, but won't the tag change?

> Changing of ImagePull signature will affect another functions

Only in 1 or 2 places; that doesn't bother me.

> seems we still need to know do we have all arches pulled or only partial of them, so return object of the function will also changed. 

If you couldn't pull all, that should be an error. To be more accurate:

* if you provided no arch, then pull everything that is available
* if you explicitly provided arches, then you must pull all of them to succeed

Reply:
> > I use p.IndexWrite(ref, im.Manifests...) inside ImagePull, so it will store all manifests
> 
> Correct, but won't the tag change?
> 

Which tag? I use the same `ref` for all arches, the difference in inside pulling of blobs for arch or skip.

> > Changing of ImagePull signature will affect another functions
> 
> Only in 1 or 2 places; that doesn't bother me.
> 

Modified, but looks it is quite a lot of changes (I put them inside the second commit), can you please check?

Reply:
Only @giggsoff can write "quite a lot of changes" but had it all done in 5 minutes.

Reply:
I see. I am thinking it through. I think there are weak points in the code design - the original, not what you are trying to fix up - which is what is making it difficult. I am trying to wrap my head around it so I can respond sanely.

Reply:
Trying to get a clearer picture of what each thing does.

* `cache.Provider.ValidateImage()`: given a reference and an architecture, ensure that the reference either points to an image for that architecture, or an index that includes an image for that architecture. Return an `lktspec.ImageSource` for that image.
* `cache.Provider.ImagePull()`: Wraps `ValidateImage()` in that if not found, pull it down. Return  an `lktspec.ImageSource` for that image.

The call chain looks something like this:

```
(lkt pkg build) Pkg.Build ---> Pkg.buildArch ----> Provider.ImagePull ----> Provider.ValidateImage
                                          ↳------------------------↑    ↑
                                          ↳----------------------------↑

(lkt build) ------------------ moby.imagePull ---> Provider.ImagePull ----> Provider.ValidateImage
```

* The ones in `lkt pkg build` use `ImagePull()` just to ensure that a particular reference includes a particular architecture. 
* The one in `lkt build` uses it to ensure it exists and get a reference to it to feed into composing the bootable OS image.

Now I (hopefully) can do a better job evaluating.

`ImagePull()` _currently_:

1. Checks the current cache for the ref and given architecture (as the image or part of the index, whatever ref points to) and returns a pointer to the specific image (unless `force`); if it fails (or if `force`):
2. Pulls down everything available in the registry (all arches)
3. Checks the cache post-pull for the ref and given architecture

When it is called by `Pkg.Build()` (in the early validation stage), it does it once, pulling the whole thing and pointing the tag at the index, but validating just for the current target architecture.

Your proposed change has the `Pkg.Build()` check (in the early validation stage), it does it for every target architecture. This is good, as it currently was checking just for the local arch. I find it surprising we didn't have a failure here before. it does so via `ImagePull()`, which, with your changes, writes the entire index each time, as retrieved from the registry, but only writes the manifests for the desired architecture? That makes sense now.

So if I call `ImagePull()` multiple times, different archs each time, I will get the same unchanged index, but a different `ImageSource` - and potentially more blobs, configs and manifests, downloaded to cache as it fills it up?

This just leaves the question of which way to do this. The "loop around `ImagePull()`" of the first commit, to the `pass multiple archs to `ImagePull()`" of the second.

Now that I understand, I am ok with it either way.  like the idea of not having to loop around it, burying that inside the func, which makes your second commit a bit cleaner, but we can go either way.

The one request I have is to add to the `ImagePull()` comment to change it from:

```go
// ImagePull takes an image name and pulls it down, writing it locally. It should be
// efficient and only write missing blobs, based on their content hash. If the ref already
// exists in the cache, it will niot pull anything, unless alwaysPull is set to true.
```

to something that describes what we have:

```go
// ImagePull takes an image name and ensures that the image manifest or index to which it refers
// exists in local cache and, if not, pulls it from the registry and writes it locally. It should be
// efficient and only write missing blobs, based on their content hash.
// It will only pull the actual blobs, config and manifest for the requested architectures, even if ref
// points to an index with multiple architectures. If the ref and all of the content for the requested
// architectures already exist in the cache, it will not pull anything, unless alwaysPull is set to true.
// If you call it multiple times, even with different architectures, the ref will continue to point to the same index.
// Only the underlying content will be added.
```

Reply:
Well, I prefer to go with the loop for now because another approach require refactoring. The code from the second commit looks overloaded.
Thank you for the comment addition, modified.

Reply:
Then we will stick with it. Comment here when you think it is ready, I will approve, and we can merge in after CI is green.

Reply:
It is ready